File: analysis/compare_ts_matlab.m
%
% compare_ts_matlab.m
%
% Tests whether mn_RPS_task.m math matches agents.ts.
% Reimplements mn_RPS_task.m probability + update equations directly,
% runs them on the same fixed 40-trial choice sequence as compare_ts_matlab.mjs,
% and writes matlab_probs.json.
%
% Run from buergi_chase/ in MATLAB:
%   cd('~/github_tasks/buergi_chase')
%   run('~/github_tasks/rps-social-cognition/analysis/compare_ts_matlab.m')
%
% Then run: python3 analysis/compare_ts_matlab.py
%

ALPHA = 0.9;
BETA  = 10;

% Payoff matrix [myAction, opponentAction], 1-indexed, R=1/P=2/S=3
% Matches comp_paymatrix output: both pi_subj and pi_bot are identical for symmetric RPS.
% Verified from comp_paymatrix.m: pay_matrix(i,j) = payoff for row player playing i vs j.
payoff = [ 0, -1,  1; ...
            1,  0, -1; ...
           -1,  1,  0];

% Softmax — matches mn_RPS_task.m softmax_fxn (no max subtraction; same result for these inputs)
softmax_fn = @(v) exp(BETA * v(:)') / sum(exp(BETA * v(:)'));  % always returns 1×3 row

% Fixed 40-trial sequence: [p_choice, bot_choice], 1-indexed (R=1, P=2, S=3)
% Identical to compare_ts_matlab.mjs SEQ.
SEQ = [ ...
  1,2; 1,3; 2,1; 3,2; 1,1; 2,3; 3,1; 1,2; 2,2; 3,3; ...
  1,3; 2,1; 3,2; 1,1; 2,3; 3,3; 1,2; 2,1; 3,3; 1,2; ...
  2,2; 3,1; 1,3; 2,2; 3,1; 1,1; 2,3; 3,2; 1,3; 2,1; ...
  3,3; 1,2; 2,1; 3,1; 1,3; 2,2; 3,3; 1,1; 2,3; 3,2; ...
];

n_trials = size(SEQ, 1);
levels   = [0, 1, 2];
results  = struct([]);

for li = 1:numel(levels)
    level  = levels(li);
    f_bot  = ones(3,1)/3;   % bot's own attraction tracker  (= agents.ts attr)
    f_subj = ones(3,1)/3;   % participant's attraction tracker (= agents.ts pAttr)

    for t = 1:n_trials
        p_choice   = SEQ(t,1);
        bot_choice = SEQ(t,2);

        % Compute probability vector BEFORE update — matches agents.ts choose() call order.
        % Equations from mn_RPS_task.m lines 86-98 (Version from CHASE paper):
        %   k=0: softmax(f_bot, β)
        %   k=1: softmax(π × f_subj, β)                         (raw f_subj, no initial softmax)
        %   k=2: bot_pred = softmax(π × f_bot, β)
        %        p = softmax(π × bot_pred', β)
        if level == 0
            p = softmax_fn(f_bot);
        elseif level == 1
            p = softmax_fn(payoff * f_subj);
        else  % level == 2
            bot_pred = softmax_fn(payoff * f_bot);
            p        = softmax_fn(payoff * bot_pred');
        end

        n = numel(results) + 1;
        results(n).trial      = t;
        results(n).level      = level;
        results(n).p_choice   = p_choice;
        results(n).bot_choice = bot_choice;
        results(n).probs      = p;
        results(n).attr       = f_bot';
        results(n).pAttr      = f_subj';

        % Deterministic update — mn_RPS_task.m lines 142-150:
        %   f_mat_subj(t,:) = (1-α)*f_mat_subj(t-1,:) + α*one_hot(choice_own(t))
        %   f_mat_bot(t,:)  = (1-α)*f_mat_bot(t-1,:)  + α*one_hot(choice_other(t))
        ind_bot           = zeros(3,1); ind_bot(bot_choice)  = 1;
        ind_subj          = zeros(3,1); ind_subj(p_choice)   = 1;
        f_bot             = (1-ALPHA)*f_bot  + ALPHA*ind_bot;
        f_subj            = (1-ALPHA)*f_subj + ALPHA*ind_subj;
    end
end

% Save as JSON
out_path = fullfile(getenv('HOME'), 'github_tasks', 'rps-social-cognition', 'analysis', 'matlab_probs.json');
json_str = jsonencode(results);
fid = fopen(out_path, 'w');
fwrite(fid, json_str, 'char');
fclose(fid);
fprintf('Written %d records → %s\n', numel(results), out_path);

File: analysis/compare_ts_matlab.mjs
/**
 * compare_ts_matlab.mjs
 *
 * Tests whether agents.ts math matches Buergi's mn_RPS_task.m math.
 * Reimplements agents.ts probability + update equations in plain JS,
 * runs them on a fixed 40-trial choice sequence, and writes ts_probs.json.
 *
 * Run:
 *   node analysis/compare_ts_matlab.mjs
 *
 * Then run compare_ts_matlab.m in MATLAB to produce matlab_probs.json,
 * then compare_ts_matlab.py to assert the two match.
 */

import { writeFileSync } from 'fs';
import { dirname, resolve } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

// Constants from config.ts
const ALPHA  = 0.9;
const BETA   = 10;
const LAMBDA = 1.0;  // bot has no loss aversion

// Payoff matrix [myAction][opponentAction], 0-indexed, R=0/P=1/S=2
// From agents.ts: payoff = [[TIE,LOSS,WIN],[WIN,TIE,LOSS],[LOSS,WIN,TIE]]
const PAYOFF = [
  [ 0, -1,  1],
  [ 1,  0, -1],
  [-1,  1,  0],
];

// Replicates agents.ts private softmax()
function softmax(v) {
  const s = v.map(x => BETA * x);
  const m = Math.max(...s);
  const e = s.map(x => Math.exp(x - m));
  const sum = e.reduce((a, b) => a + b, 0);
  return e.map(x => x / sum);
}

// Replicates agents.ts private ev() — with lambda on negative payoffs
function ev(dist) {
  return PAYOFF.map(row =>
    row.reduce((s, v, j) => s + (v < 0 ? v * LAMBDA : v) * dist[j], 0)
  );
}

// Replicates agents.ts private probs() for a given level
function probs(attr, pAttr, level) {
  if (level === 0) return softmax(attr);
  const seed = level % 2 === 0 ? attr : pAttr;
  let p = softmax(ev(seed));
  for (let k = 1; k < level; k++) {
    p = softmax(ev(p));
  }
  return p;
}

// Fixed 40-trial sequence: [p_choice, bot_choice], 1-indexed (R=1, P=2, S=3)
// Shared with compare_ts_matlab.m — must be identical.
const SEQ = [
  [1,2],[1,3],[2,1],[3,2],[1,1],[2,3],[3,1],[1,2],[2,2],[3,3],
  [1,3],[2,1],[3,2],[1,1],[2,3],[3,3],[1,2],[2,1],[3,3],[1,2],
  [2,2],[3,1],[1,3],[2,2],[3,1],[1,1],[2,3],[3,2],[1,3],[2,1],
  [3,3],[1,2],[2,1],[3,1],[1,3],[2,2],[3,3],[1,1],[2,3],[3,2],
];

const results = [];

for (const level of [0, 1, 2]) {
  let attr  = [1/3, 1/3, 1/3];
  let pAttr = [1/3, 1/3, 1/3];

  for (let t = 0; t < SEQ.length; t++) {
    const [pChoice, botChoice] = SEQ[t];

    // Record pre-update state and probability vector
    results.push({
      trial:      t + 1,
      level,
      p_choice:   pChoice,
      bot_choice: botChoice,
      probs:      probs(attr, pAttr, level),
      attr:       [...attr],
      pAttr:      [...pAttr],
    });

    // Replicates agents.ts update(): 1-indexed choices → 0-indexed
    const botIdx = botChoice - 1;
    const pIdx   = pChoice   - 1;
    attr  = attr.map( (a, i) => a + ALPHA * ((i === botIdx ? 1 : 0) - a));
    pAttr = pAttr.map((a, i) => a + ALPHA * ((i === pIdx   ? 1 : 0) - a));
  }
}

const outPath = resolve(__dirname, 'ts_probs.json');
writeFileSync(outPath, JSON.stringify(results, null, 2));
console.log(`Written ${results.length} records → ${outPath}`);
